cmd/clusterctl/client/cluster/internal/dryrun/client.go

```diff
@@ -0,0 +1,354 @@
+	// make sure that the object is available in the fake client.
+	if err := c.ensureObjInFakeClient(ctx, obj); err != nil {
+		return errors.Wrap(err, "failed to ensure object is fake client")
+	}
+	err := c.fakeClient.Patch(ctx, obj, patch, opts...)
+	if err == nil {
+		key := trackerIDOf(obj)
+		// Check if the key already exists in the updated tracker. If not add
+		// the key and capture the original object.
+		// The original object will be used to calculate the effective changes
+		// to the object.
+		if _, ok := c.changeTracker.patched[key]; !ok {
+			c.changeTracker.patched[key] = originalObj
+		}
+	}
+	return err
+}
+
+// DeleteAllOf deletes all objects of the given type matching the given options.
+// NOTE: Topology reconciler does not use DeleteAllOf, so we are skipping implementation for now.
+func (c *Client) DeleteAllOf(ctx context.Context, obj client.Object, opts ...client.DeleteAllOfOption) error {
+	panic("DeleteAllOf method is not supported by the dryrun client")
+}
+
+// Status returns a client which can update status subresource for kubernetes objects.
+func (c *Client) Status() client.StatusWriter {
+	return c.fakeClient.Status()
+}
+
+// Scheme returns the scheme this client is using.
+func (c *Client) Scheme() *runtime.Scheme {
+	return c.fakeClient.Scheme()
+}
+
+// RESTMapper returns the rest this client is using.
+func (c *Client) RESTMapper() meta.RESTMapper {
+	return c.fakeClient.RESTMapper()
+}
+
+// Changes generates a summary of all the changes observed from the creation of the dry run client
+// to when this function is called.
+func (c *Client) Changes(ctx context.Context) (*ChangeSummary, error) {
+	changes := &ChangeSummary{
+		Created:  []*unstructured.Unstructured{},
+		Modified: []*PatchSummary{},
+		Deleted:  []*unstructured.Unstructured{},
+	}
+
+	// Handle created objects.
+	for id := range c.changeTracker.created {
+		if _, ok := c.changeTracker.deleted[id]; ok {
+			// If the object is created and deleted withting the same reconcile loop
+			// then we do not need to track it.
+			continue
+		}
+		obj := &unstructured.Unstructured{}
+		obj.SetGroupVersionKind(id.gvk)
+		obj.SetNamespace(id.key.Namespace)
+		obj.SetName(id.key.Name)
+		if err := c.fakeClient.Get(ctx, id.key, obj); err != nil {
+			return nil, errors.Wrapf(err, "failed to read created object %s", id.key.String())
+		}
+		changes.Created = append(changes.Created, obj)
+	}
+
+	// Handle patched objects.
+	// If an object is updated multiple times, all the changes are flattend. Only the initial
+	// and final states are captured.
+	for id, original := range c.changeTracker.patched {
+		if _, ok := c.changeTracker.created[id]; ok {
+			// If an object is create and modified in the same reconcile loop
+			// we only track it under 'created'.
+			continue
+		}
+		if _, ok := c.changeTracker.deleted[id]; ok {
+			// If an object is modified and delete in the same reconcile loop
+			// we only track it under 'deleted'.
+			continue
+		}
```

Comment:
This logic works given the current behavior of the topology reconciler, but it could fail when the order of the operations is different.
We should get this into create, patch, delete

Reply:
Oh yes. I see what you mean. 
Example: If an object is created followed by deleted then we should ignore the object. However, if the object is deleted and then re-created (may be with some changes) we should show it under updated(?)(or re-created? 🤔).

Reply:
If the client would have to replicate reality very closely it depends on the order :) (i.e. it should be the same as in reality)

I think if we add a comment we can simplify by assuming the topology reconciler never creates and object after it deleted it before (which is imho 100% safe given the implementation of reconcile state)